[CIR][CIRGen][TBAA] Initial TBAA support

PikachuHyA opened the pull request:
This is the first patch to support TBAA, following the discussion at https://github.com/llvm/clangir/pull/1076#discussion_r1835031415

- add skeleton for CIRGen, utilizing `decorateOperationWithTBAA`
- add empty implementation in `CIRGenTBAA`
- introduce `CIR_TBAAAttr` with empty body
- attach `CIR_TBAAAttr` to `LoadOp` and `StoreOp`
- no handling of vtable pointer
- no LLVM lowering

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -596,6 +597,7 @@ def LoadOp : CIR_Op<"load", [
     (`volatile` $is_volatile^)?
     (`align` `(` $alignment^ `)`)?
     (`atomic` `(` $mem_order^ `)`)?
+    (`tbaa` `(` $tbaa^ `)`)?
```
Given tbaa might be a bit more noisy (it's an array), let's move it after the addr type in load and after value type in loads, how about that? Anywhere after the actual params look fine to me. Can you also add alias printing for TBAAAttr?

Also, because this is an optional array, I don't think we'll be able to add alias printing for the full thing, but only for each element, so instead of `cir.load ... tbaa(#tbaa_array20)` we would have `cir.load ... tbaa([#tbaa1, #tbaa2, ...])`, which could be fine too? Have you considered having one `tbaa_array` attribute that holds an array of `tbaa_attrs`? We don't necessarily need to go this way, but just curious if you have extra ideas.

PikachuHyA replied:
> Given tbaa might be a bit more noisy (it's an array), let's move it after the addr type in load and after value type in loads, how about that?

I agree.
Moving the TBAA information to the end improves readability of the CIR code, as the TBAA alias tags can be significantly longer than other components, such as operation names and operands. Additionally, this adjustment is more friendly for other unit tests that do not need to test TBAA. We do not need to modify the code for these tests. for examples,`clang/test/CIR/CodeGen/binop.cpp`, `clang/test/CIR/CodeGen/dynamic-cast-exact.cpp`, `clang/test/CIR/CodeGen/ternary.cpp`, and `clang/test/CIR/Transforms/ternary-fold.cpp`.

PikachuHyA replied:
Regarding the design of the TBAA attribute, I have provided a simple explanation https://github.com/llvm/clangir/pull/1116#issuecomment-2482792967

Overall, we need to have CIR TBAA attributes that correspond to the LLVM TBAA attributes, such as `TBAAMember`, `TBAATag` and `TBAATypeDescriptor`. I prefer that we discuss the specific design and implementation details of the TBAA attributes in the next patch after this patch is landed.

bcardosolopes commented on clang/test/CIR/CodeGen/ternary.cpp:
```diff
@@ -9,7 +9,7 @@ int x(int y) {
 // CHECK:     %1 = cir.alloca !s32i, !cir.ptr<!s32i>, ["__retval"] {alignment = 4 : i64}
 // CHECK:     cir.store %arg0, %0 : !s32i, !cir.ptr<!s32i>
-// CHECK:     %2 = cir.load %0 : !cir.ptr<!s32i>, !s32i
+// CHECK:     %2 = cir.load tbaa({{.*}}) %0 : !cir.ptr<!s32i>, !s32i
```
Follow up from previous comment: after you add the alias we should filecheck the alias, even if it's empty for now, instead of `{{.*}}` here and elsewhere.

bcardosolopes reviewed (changes requested):
Thanks a bunch for breaking this up, mostly LGTM with few nits

seven-mile commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -3511,8 +3519,71 @@ void CIRGenModule::emitGlobalAnnotations() {
+    return nullptr;
+  }
+  auto ret = tbaa->getAccessTagInfo(tbaaInfo);
+  llvm::errs() << "ret: " << ret << "\n";
```
Debug print ; )

PikachuHyA replied:
Good catch!

seven-mile reviewed (commented):
Cool! Looking forward to it~

PikachuHyA commented:
### Summary of Changes

- Added alias printing for `CIR_TBAAAttr`.
- Moved TBAA printing to the end of the output.
- Unified variable names to use camelCase.
- Revised comments for clarity.
- Removed unnecessary debug output.

### Explanations

- **Why is the type of TBAA `mlir::ArrayAttr`?**  
  The introduction of the `tbaa` attribute as `OptionalAttr<SymbolRefArrayAttr>` is in the commit https://github.com/llvm/llvm-project/commit/edfefd7fc50b995c4dfd81d5f9db53c1a9e11580 . The follow-up changes made in https://github.com/llvm/llvm-project/commit/1dda134f85d4fab7ace7b9ec9b57a0319c14de8f further address this design choice, which changes `tbaa` attribute as `OptionalAttr<LLVM_TBAATagArrayAttr>`. 

  The rationale behind this decision stems from the following comment:
  
  ```cpp
  // LLVM IR currently does not support attaching more than one TBAA access tag
  // to a memory accessing instruction. While supporting this in the future may
  // be beneficial, for now we simply ignore the metadata if the MLIR operation
  // has multiple access tags.
  ```
  
  For additional context, please refer to [ModuleTranslation.cpp#L1101-L1109](https://github.com/llvm/llvm-project/blob/edfefd7fc50b995c4dfd81d5f9db53c1a9e11580/mlir/lib/Target/LLVMIR/ModuleTranslation.cpp#L1101-L1109).

  In this patch, we use `mlir::ArrayAttr` to simplify the TBAA attribute, allowing for a quicker landing of the patch. We will refine the TBAA attribute design in the next patch.

bcardosolopes commented:
> * In this patch, we use `mlir::ArrayAttr` to simplify the TBAA attribute, allowing for a quicker landing of the patch. We will refine the TBAA attribute design in the next patch.

I dislike optional<ArrayAttr> because there are two levels on null checking here (the optional could be empty and the array could also be empty), which is quite confusing. Better yet that LLVM doesn't really support multiple entries. I suggested to you in previous comments to create a new attr if necessary to wrap the array exactly for this reason: the array becomes an implementation detail and it doesn't really matter if we remove the array later, since it's internal state of the attribute.

Anyways, I'm fine if this comes as incremental work though. Thanks
